circuits/bigint.circom

```diff
@@ -0,0 +1,566 @@
+        ands[i] = AND();
+        eq_ands[i] = AND();
+        ors[i] = OR();
+
+        if (i == k - 2) {
+           ands[i].a <== eq[k - 1].out;
+           ands[i].b <== lt[k - 2].out;
+           eq_ands[i].a <== eq[k - 1].out;
+           eq_ands[i].b <== eq[k - 2].out;
+           ors[i].a <== lt[k - 1].out;
+           ors[i].b <== ands[i].out;
+        } else {
+           ands[i].a <== eq_ands[i + 1].out;
+           ands[i].b <== lt[i].out;
+           eq_ands[i].a <== eq_ands[i + 1].out;
+           eq_ands[i].b <== eq[i].out;
+           ors[i].a <== ors[i + 1].out;
+           ors[i].b <== ands[i].out;
+        }
+     }
+     out <== ors[0].out;
+}
+
+template BigIsEqual(k){
+    signal input in[2][k];
+    signal output out;
+    component isEqual[k+1];
+    var sum = 0;
+    for(var i = 0; i < k; i++){
+        isEqual[i] = IsEqual();
+        isEqual[i].in[0] <== in[0][i];
+        isEqual[i].in[1] <== in[1][i];
+        sum = sum + isEqual[i].out;
+    }
+
+    isEqual[k] = IsEqual();
+    isEqual[k].in[0] <== sum;
+    isEqual[k].in[1] <== k;
+    out <== isEqual[k].out;
+}
+
+// leading register of b should be non-zero
+template BigMod(n, k) {
+    assert(n <= 126);
+    signal input a[2 * k];
+    signal input b[k];
+
+    signal output div[k + 1];
+    signal output mod[k];
+
+    var longdiv[2][100] = long_div(n, k, k, a, b);
+    for (var i = 0; i < k; i++) {
+        div[i] <-- longdiv[0][i];
+        mod[i] <-- longdiv[1][i];
+    }
+    div[k] <-- longdiv[0][k];
+    component div_range_checks[k + 1];
+    for (var i = 0; i <= k; i++) {
+        div_range_checks[i] = Num2Bits(n);
+        div_range_checks[i].in <== div[i];
+    }
+    component mod_range_checks[k];
+    for (var i = 0; i < k; i++) {
+        mod_range_checks[i] = Num2Bits(n);
+        mod_range_checks[i].in <== mod[i];
+    }
+
+    component mul = BigMult(n, k + 1);
+    for (var i = 0; i < k; i++) {
+        mul.a[i] <== div[i];
+        mul.b[i] <== b[i];
+    }
+    mul.a[k] <== div[k];
+    mul.b[k] <== 0;
+
+    component add = BigAdd(n, 2 * k + 2);
+    for (var i = 0; i < 2 * k; i++) {
+        add.a[i] <== mul.out[i];
+        if (i < k) {
+            add.b[i] <== mod[i];
+        } else {
+            add.b[i] <== 0;
+        }
+    }
+    add.a[2 * k] <== mul.out[2 * k];
+    add.a[2 * k + 1] <== mul.out[2 * k + 1];
+    add.b[2 * k] <== 0;
+    add.b[2 * k + 1] <== 0;
+
+    for (var i = 0; i < 2 * k; i++) {
+        add.out[i] === a[i];
+    }
+    add.out[2 * k] === 0;
+    add.out[2 * k + 1] === 0;
+
+    component lt = BigLessThan(n, k);
+    for (var i = 0; i < k; i++) {
+        lt.a[i] <== mod[i];
+        lt.b[i] <== b[i];
+    }
+    lt.out === 1;
+}
+
+// a[i], b[i] in 0... 2**n-1
+// represent a = a[0] + a[1] * 2**n + .. + a[k - 1] * 2**(n * k)
+// assume a >= b
+template BigSub(n, k) {
+    assert(n <= 252);
+    signal input a[k];
+    signal input b[k];
+    signal output out[k];
+    signal output underflow;
```

Comment:
this circuit assumes that a>=b, then `underflow` isn't required since it's always 0.